Cip tool

ChemMitch opened the pull request:
Select method of chiral center labeling based on estimated complexity of molecule.
2 externally-settable parameters govern this estimation.

ChemMitch committed:
new unit test

ChemMitch committed:
experimentation on how to speed up structure CIP analysis

ChemMitch committed:
minor tweak

ChemMitch committed:
testing out an idea to speed up chirality calculations for
high-symmetric ring system

ChemMitch committed:
tentative Idea for speeding up labelling.
Will be tested

ChemMitch committed:
small change

ChemMitch committed:
experimentation

ChemMitch committed:
moved determination of which labelling method to use to CdkChemicalImpl

ChemMitch committed:
tweak to unit test

ChemMitch committed:
removed a lot of debug logging and streamlined getTetrahedrals

ChemMitch committed:
Added parameters to set complexityCutoff and maxUndefined

ChemMitch committed:
adding data (mol) files for unit tests

ChemMitch committed:
removed much of the System.out.print* statements
tweak to testOneMol: explicitly set maxUndefined and complexityCutoff

ChemMitch committed:
implemented applyParameters in CdkChemical2FactoryImpl

ChemMitch committed:
moved implementation of applyParameters to factory class

ChemMitch committed:
changed name of method

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -130,6 +133,40 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
+		catch (NumberFormatException ignore){
+		}
+	}
+	public static Integer getMaxUndefined() {
```
Lets rename this to include stereo

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -130,6 +133,40 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
+		CdkChemicalImpl.maxUndefined = maxUndefined;
+	}
+
+	public void setMaxUndefined(String maxUndefined) {
```
We shouldn't have this setter, only the actual Integer type one, I think.

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -130,6 +133,40 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
+		CdkChemicalImpl.complexityCutoff = complexityCutoff;
+	}
+
+	public static void setComplexityCutoff(String complexityCutoff) {
```
Let's remove this one.

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -130,6 +133,40 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
 	private static IChemObjectBuilder CHEM_OBJECT_BUILDER = SilentChemObjectBuilder.getInstance();
 	private CDKHydrogenAdder hydrogenAdder;
+
+	private static int complexityCutoff = 7;
```
Add a comment above this

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -130,6 +133,40 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
+		} catch (NumberFormatException ignore){}
+	}
+
+	private static int maxUndefined = 20;
```
Add a comment above this.

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -140,9 +177,10 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
-	
 
-    static{
+	//private boolean hasBeenLabeled = false;
```
remove

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -187,21 +225,26 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
-    CachedSupplier<Void> cahnIngoldPrelogSupplier = CachedSupplier.of(()->{
+
+
+	private Boolean isComplex = false;
```
remove this

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -187,21 +225,26 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
+	private Boolean isComplex = false;
+
+    CachedSupplier<Integer> cahnIngoldPrelogSupplier = CachedSupplier.of(()->{
+		this.isComplex = complexitySupplier.get();
```
make a local variable

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -338,23 +391,24 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
 	    	}
-	    	
+
+			System.out.printf("number of potential centers: %d; number of undefined center: %d\n", potentialSet.size(),
```
change to log or remove

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -213,10 +256,20 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
+				//if( !hasBeenLabeled) {
+				//	hasBeenLabeled = true;
+					if( isComplex) {
+						System.out.println("This molecule is considered complex");
```
change to log or remove

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalInchiImplFactory.java:
```diff
@@ -87,7 +89,7 @@ private static boolean isAtomThatMessesUpInchi(Atom atom) {
 	@Override
 	public InChiResult asStdInchi(Chemical chemical, boolean trustCoordinates) throws IOException {
-		
+		System.out.println("in asStdInchi");
```
remove

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalInchiImplFactory.java:
```diff
@@ -54,8 +54,10 @@ public class CdkChemicalInchiImplFactory implements InchiImplFactory{
+		System.out.println("in CdkChemicalInchiImplFactory ctor");
 		try {
 			factory = InChIGeneratorFactory.getInstance();
+			System.out.printf("got factor: %s\n", factory.getClass().getName());
```
remove

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalInchiImplFactory.java:
```diff
@@ -54,8 +54,10 @@ public class CdkChemicalInchiImplFactory implements InchiImplFactory{
 	//CDK by default turns of auxInfo
 	private static InchiOptions MOLWITCH_INCHI_OPTIONS = new InchiOptions.InchiOptionsBuilder().build();
 	public CdkChemicalInchiImplFactory() {
+		System.out.println("in CdkChemicalInchiImplFactory ctor");
```
remove

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalInchiImplFactory.java:
```diff
@@ -134,6 +136,7 @@ private InChiResult.Status toChemkitStatus(InchiStatus returnStatus) {
 	}
 	@Override
 	public Chemical parseInchi(String inchi) throws IOException {
+		System.out.println("in CdkChemicalInchiImplFactory.parseInchi");
```
remove

tylerperyeaFDA commented on src/main/java/org/openscience/cdk/geometry/cip/CIPToolMod.java:
```diff
@@ -111,21 +124,33 @@
-    		return;
-    	}else {
+    public static void label(IAtomContainer container, CdkChemicalImpl chemical) {
+        totalCallsToLabel++;
```
remove

tylerperyeaFDA commented on src/main/java/org/openscience/cdk/geometry/cip/CIPToolMod.java:
```diff
@@ -271,8 +296,61 @@ private static int permParity(final ILigand[] ligands) {
+                maxRings = currentRingTotal;
+            }
+        }
+        System.out.printf("total number of fragments for this chemical: %d\n", fragmentCount);
```
remove system.out stuff here

tylerperyeaFDA commented on src/main/java/org/openscience/cdk/geometry/cip/CIPToolMod.java:
```diff
@@ -271,8 +296,61 @@ private static int permParity(final ILigand[] ligands) {
+        return maxRings;
+    }
+
+    private void writeMolTemp(CdkChemicalImpl chem) {
```
this method shouldn't be needed ... can we remove?

tylerperyeaFDA commented on src/test/java/TestChiralRead.java:
```diff
@@ -1268,5 +1313,646 @@ public void testQueryStructureSimpleTetrahedralStereoDoesntError() throws Except
+	}
+
+	@Test
+	public void testSlowChiralityTrueMultiple() throws IOException {
```
I really like how this was done.

tylerperyeaFDA commented on src/test/java/TestChiralRead.java:
```diff
@@ -1268,5 +1313,646 @@ public void testQueryStructureSimpleTetrahedralStereoDoesntError() throws Except
+
+	private boolean testOneMol(TestMol testMol) throws IOException {
+		String userDirectory = new File("").getAbsolutePath();
+		String fileName =userDirectory + "/src/test/resources/mols/" + testMol.molfileName;
```
In general I think this can get a bit dodgy based on what OS/classloaders/etc are being used. I think the more kosher way to do it is like this:

```
 String molfileText = IOUtils.toString(
      this.getClass().getResourceAsStream("mols/" + testMol.molfileName),
      "UTF-8"
    );
    Chemical c1=Chemical.parse(molfileText );
    
```

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemical2FactoryImpl.java:
```diff
@@ -529,4 +530,15 @@ public boolean isDefault() {
 	}
 
+	public static void applyParameters(Map<String, Object> params){
+		System.out.println("in CdkChemical2FactoryImpl.applyParameters");
```
remove debug lines

tylerperyeaFDA commented on src/main/java/org/openscience/cdk/geometry/cip/CIPToolMod.java:
```diff
@@ -111,21 +124,33 @@
-	
-	/**
+
+    private final static int MAX_RINGS = 5;
```
Remove this and the following line, they're no longer needed

tylerperyeaFDA commented on src/main/java/org/openscience/cdk/geometry/cip/CIPToolMod.java:
```diff
@@ -271,8 +296,61 @@ private static int permParity(final ILigand[] ligands) {
+                copy.removeBond(i);
+            }
+        }
+
```
If we didn't find a ring bond by here, we should just return 0 I think, save some unnecessary processing.

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -2438,8 +2500,9 @@ public void remove() {
-	
 
+	@Override
+	public Class getFactoryClass() {
```
I'm 99.9% sure we won't need this. We can get the factory from the ImplUtil class if needed, and I don't think we need to go from a ChemicalImpl to its factory? I'd rather not add such a method unless we're sure we need it.

tylerperyeaFDA reviewed (changes requested):
Mostly I'd just like to remove some debug stuff and some older stuff we tried but ended up going a different way. The largest ring detection should probably have a little short-circuit so it doesn't always do the full ring analysis if there's no ring detected.

I also think we can get away without the getFactoryClass method.

Other than that it looks good!

ChemMitch committed:
removed debug output statements
Removed unnecessary methods and parameters, per reviewer
Misc. clean-ups

ChemMitch committed:
clean-up

ChemMitch committed:
removed unused property

ChemMitch committed:
Made test use realistic parameters.
Added labels on uncertain stereocenters

ChemMitch committed:
updated procesing of bonds

ChemMitch committed:
made ringsSearcherSupplier resistent to exceptions thrown by AllRingsFinder

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemical2FactoryImpl.java:
```diff
@@ -529,4 +530,16 @@ public boolean isDefault() {
+			CdkChemicalImpl.setComplexityCutoff((Integer) params.get("complexityCutoff"));
+			System.out.printf("complexityCutoff: %s\n", params.get("complexityCutoff"));
+		}
+		if( params.get("maxUndefined") != null ) {
```
Can we rename to maxUndefinedStereo ?

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -130,6 +134,30 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
+		CdkChemicalImpl.maxUndefinedStereoCenters = maxUndefinedStereoCenters;
+	}
+
+	private static int maxUndefinedStereoCenters = 20;
```
Add comment above

tylerperyeaFDA replied:
And default to like ... 5? It should usually be small

tylerperyeaFDA commented on src/main/java/gov/nih/ncats/molwitch/cdk/CdkChemicalImpl.java:
```diff
@@ -187,21 +214,25 @@ public class CdkChemicalImpl implements ChemicalImpl<CdkChemicalImpl>{
 
-    CachedSupplier<Void> perceiveAtomTypesOfNonQueryAtoms = CachedSupplier.of(()->{
+	CachedSupplier<Boolean> complexitySupplier =CachedSupplier.of(()->{
+		Logger.getLogger(this.getClass().getName()).info("getSizeOfLargestRingSystem(this): " + getSizeOfLargestRingSystem(this));
```
Remove this line, or compute once and reference the field. Right now the getSizeoFLargestRingSystem is called twice

tylerperyeaFDA commented on src/main/java/org/openscience/cdk/geometry/cip/CIPToolMod.java:
```diff
@@ -271,8 +287,45 @@ private static int permParity(final ILigand[] ligands) {
+            Atom atom = copy.getAtom(i);
+            if(atom.getBondCount() == 0) {
+                //System.out.printf("atom %d of symbol %s has no bonds and will be deleted\n", i, atom.getSymbol());
+                copy.removeAtom(i);
```
Let's do same trick here to add to a list and tehm remove them all. I think otherwise there are some other caches that can go wrong.

ChemMitch committed:
renamed maxUndefined to maxUndefinedStereoCenters
cleaned up atom deletion in getSizeOfLargestRingSystem
prevent double calls to getSizeOfLargestRingSystem
use real logging instead of println
